app/src/parser_impl.c

```diff
@@ -0,0 +1,873 @@
+}
+
+parser_error_t _readAccounts(parser_context_t *c, uint8_t accounts[][32], size_t *numAccounts, size_t maxAccounts)
+{
+    CHECK_ERROR(_readArraySize(c, numAccounts))
+
+    if (*numAccounts > maxAccounts) {
+        return parser_msgpack_array_too_big;
+    }
+
+    for (size_t i = 0; i < *numAccounts; i++) {
+        CHECK_ERROR(_readBinFixed(c, accounts[i], sizeof(accounts[0])))
+    }
+
+    return parser_ok;
+}
+
+parser_error_t _readStateSchema(parser_context_t *c, state_schema *schema)
+{
+    size_t mapSize = 0;
+    CHECK_ERROR(_readMapSize(c, &mapSize))
+    uint8_t key[32];
+    for (size_t i = 0; i < mapSize; i++) {
+        CHECK_ERROR(_readString(c, key, sizeof(key)))
+        if (strcmp((char*)key, "nui") == 0) {
+            CHECK_ERROR(_readInteger(c, &schema->num_uint))
+        } else if (strcmp((char*)key, "nbs") == 0) {
+            CHECK_ERROR(_readInteger(c, &schema->num_byteslice))
+        } else {
+            return parser_msgpack_unexpected_key;
+        }
+    }
+
+    return parser_ok;
+}
+
+parser_error_t _readArrayU64(parser_context_t *c, uint64_t elements[], size_t *numElements, size_t maxElements)
+{
+    CHECK_ERROR(_readArraySize(c, numElements))
+
+    if (*numElements > maxElements) {
+        return parser_msgpack_array_too_big;
+    }
+
+    for (size_t i = 0; i < *numElements; i++) {
+        CHECK_ERROR(_readInteger(c, &elements[i]))
+    }
+
+    return parser_ok;
+}
+
+static parser_error_t _readTxType(parser_context_t *c, parser_tx_t *v)
+{
+    uint8_t buff[100] = {0};
+    uint8_t buffLen = sizeof(buff);
+    uint8_t typeStr[50] = {0};
+    uint8_t typeStrLen = sizeof(typeStr);
+    uint16_t currentOffset = c->offset;
+    c->offset = 0;
+
+    for (size_t offset = 0; offset < c->bufferLen; offset++) {
+        if (_readString(c, buff, buffLen) == parser_ok) {
+            if (strcmp((char*)buff, KEY_COMMON_TYPE) == 0) {
+                if (_readString(c, typeStr, typeStrLen) == parser_ok) {
+                    if (strcmp((char *) typeStr, KEY_TX_PAY) == 0) {
+                        v->type = TX_PAYMENT;
+                        break;
+                    } else if (strcmp((char *) typeStr, KEY_TX_KEYREG) == 0) {
+                        v->type = TX_KEYREG;
+                        break;
+                    } else if (strcmp((char *) typeStr, KEY_TX_ASSET_XFER) == 0) {
+                        v->type = TX_ASSET_XFER;
+                        break;
+                    } else if (strcmp((char *) typeStr, KEY_TX_ASSET_FREEZE) == 0) {
+                        v->type = TX_ASSET_FREEZE;
+                        break;
+                    } else if (strcmp((char *) typeStr, KEY_TX_ASSET_CONFIG) == 0) {
+                        v->type = TX_ASSET_CONFIG;
+                        break;
+                    } else if (strcmp((char *) typeStr, KEY_TX_APPLICATION) == 0) {
+                        v->type = TX_APPLICATION;
+                        break;
+                    }
```

Comment:
let's extract this into another function. it is fine if you inline it later. But this level of nesting makes thing difficult to review and test